onair/data_handling/sbn_adapter.py

```diff
@@ -0,0 +1,198 @@
+            if "message_headers" in str(field_type) and hasattr(field_type, "_fields_"):
+                for sub_field_name, sub_field_type in field_type._fields_:
+                    gather_field_names_helper(field_name + "." + sub_field_name, sub_field_type,field_names)
+            else:
+                field_names.append(field_name)
+
+        field_names = []
+        gather_field_names_helper(field_name, field_type, field_names)
+        return field_names
+
+    def parse_meta_data_file(self, meta_data_file, ss_breakdown):
+        self.msgID_lookup_table = {}
+        self.currentData = []
+
+        # pull out message ids
+        file = open(meta_data_file, 'rb')
+        file_str = file.read()
+
+        meta_config = json.loads(file_str)
+        file.close()
+
+        if 'channels' not in meta_config.keys():
+            raise ConfigKeyError(f'Config file: \'{meta_data_file}\' ' \
+                                  'missing required key \'channels\'')
+
+        # Copy message ID table from .json, convert string hex to ints for ID
+        for key in meta_config['channels']:
+            self.msgID_lookup_table[int(key, 16)] = meta_config['channels'][key]
+
+        # Use eval() to convert class name from .json to match with message_headers.py
+        for key in self.msgID_lookup_table:
+            msg_struct_name = self.msgID_lookup_table[key][1]
+            self.msgID_lookup_table[key][1] = eval("msg_hdr." + msg_struct_name)
+
+        # populate headers and reserve space for data
+        for x in range(0,2):
+            self.currentData.append({'headers':[], 'data':[]})
+
+            for msgID in self.msgID_lookup_table.keys():
+                app_name, data_struct = self.msgID_lookup_table[msgID]
+                struct_name = data_struct.__name__
+                # Skip the header, walk through the stuct
+                for field_name, field_type in data_struct._fields_[1:]:
+                    field_names = self.gather_field_names(app_name + "." + field_name, field_type)
+                    for field_name in field_names:
+                        self.currentData[x]['headers'].append(field_name)
+                        self.currentData[x]['data'].append([0]) #initialize all the data arrays with zero
+
+        return extract_meta_data_handle_ss_breakdown(meta_data_file, ss_breakdown)
+
+    def process_data_file(self, data_file):
+        print("SBN Adapter ignoring data file (telemetry should be live)")
+
+    def get_vehicle_metadata(self):
+        return self.all_headers, self.binning_configs['test_assignments']
+
+    def get_next(self):
+        """Provides the latest data from SBN in a dictionary of lists structure.
+        Returned data is safe to use until the next get_next call.
+        Blocks until new data is available."""
+
+        data_available = False
+
+        while not data_available:
+            with self.new_data_lock:
+                data_available = self.has_data()
```

Comment:
Why is the has_data() function needed here instead of just checking self.new_data? Is this needed to ease unit testing?

Reply:
Yes this change was made to have more control during unit testing. Additionally, I saw that redis_adapter had also used a has_data() function instead of just checking self.new_data.